packages/ckeditor5-html-support/tests/emptyblocks.js

```diff
@@ -0,0 +1,331 @@
+		it( 'should set htmlEmptyBlock attribute on empty table cell', () => {
+			editor.setData( '<figure class="table"><table><tbody><tr><td></td></tr></tbody></table></figure>' );
+
+			expect( getModelData( model, { withoutSelection: true } ) ).to.equal(
+				'<table><tableRow><tableCell htmlEmptyBlock="true"><paragraph></paragraph></tableCell></tableRow></table>'
+			);
+		} );
+
+		it( 'should preserve empty table cells in data output', () => {
+			editor.setData( '<figure class="table"><table><tbody><tr><td></td><td>foo</td><td></td></tr></tbody></table></figure>' );
+
+			expect( editor.getData() ).to.equal(
+				'<figure class="table"><table><tbody><tr><td></td><td>foo</td><td></td></tr></tbody></table></figure>'
+			);
+		} );
+
+		it( 'should preserve empty cells mixed with non-empty ones', () => {
+			editor.setData( '<figure class="table"><table><tbody><tr><td>foo</td><td></td><td>bar</td></tr></tbody></table></figure>' );
+
+			expect( getModelData( model, { withoutSelection: true } ) ).to.equal(
+				'<table>' +
+					'<tableRow>' +
+						'<tableCell><paragraph>foo</paragraph></tableCell>' +
+						'<tableCell htmlEmptyBlock="true"><paragraph></paragraph></tableCell>' +
+						'<tableCell><paragraph>bar</paragraph></tableCell>' +
+					'</tableRow>' +
+				'</table>'
+			);
+
+			expect( editor.getData() ).to.equal(
+				'<figure class="table"><table><tbody><tr><td>foo</td><td></td><td>bar</td></tr></tbody></table></figure>'
+			);
+		} );
+
+		it( 'should set htmlEmptyBlock attribute on table cell with whitespace', () => {
+			editor.setData( '<figure class="table"><table><tbody><tr><td> </td></tr></tbody></table></figure>' );
+
+			expect( editor.getData() ).to.equal(
+				'<figure class="table"><table><tbody><tr><td></td></tr></tbody></table></figure>'
+			);
+		} );
+
+		it( 'should not add auto paragraphs to empty table cells with htmlEmptyBlock', () => {
+			editor.setData( '<figure class="table"><table><tbody><tr><td></td></tr></tbody></table></figure>' );
+
+			expect( getModelData( model, { withoutSelection: true } ) ).to.equal(
+				'<table>' +
+					'<tableRow>' +
+						'<tableCell htmlEmptyBlock="true"><paragraph></paragraph></tableCell>' +
+					'</tableRow>' +
+				'</table>'
+			);
+
+			expect( editor.getData() ).to.equal(
+				'<figure class="table"><table><tbody><tr><td></td></tr></tbody></table></figure>'
+			);
+		} );
+	} );
+
+	describe( 'lists integration', () => {
+		it( 'should set htmlEmptyBlock on empty list item', () => {
+			editor.setData( '<paragraph>A</paragraph><ul><li></li></ul>' );
+
+			const modelData = getModelData( model, { withoutSelection: true } );
+			const normalizedData = modelData.replace( / listItemId="[^"]+"/g, '' );
+
+			expect( normalizedData ).to.equal(
+				'<paragraph>A</paragraph>' +
+				'<paragraph htmlEmptyBlock="true" listIndent="0" listType="bulleted"></paragraph>'
+			);
+
+			expect( editor.getData() ).to.equal( '<p>A</p><ul><li><p></p></li></ul>' );
```

Comment:
I'd expect the output would not have an empty `<p>` as there wasn't one in the input.

Reply:
Done https://github.com/ckeditor/ckeditor5/pull/17756/commits/99f91c7f42db292e85acf85b9ffb68ab520927e6